examples/ota-provider-app/esp32/main/main.cpp

```diff
@@ -0,0 +1,232 @@
+#include "esp_spi_flash.h"
+#include "esp_system.h"
+#include "esp_wifi.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "nvs_flash.h"
+#include <app-common/zap-generated/callback.h>
+#include <app/server/Server.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+#include <cmath>
+#include <cstdio>
+#include <string>
+#include <vector>
+
+#include <credentials/DeviceAttestationCredsProvider.h>
+#include <credentials/examples/DeviceAttestationCredsExample.h>
+
+#include <lib/support/ErrorStr.h>
+
+#include <app/clusters/ota-provider/ota-provider.h>
+#include <ota-provider-common/BdxOtaSender.h>
+#include <ota-provider-common/OTAProviderExample.h>
+
+using chip::BitFlags;
+using chip::app::Clusters::OTAProviderDelegate;
+using chip::bdx::TransferControlFlags;
+using chip::Callback::Callback;
+using chip::Messaging::ExchangeManager;
+using namespace ::chip;
+using namespace ::chip::System;
+using namespace ::chip::Credentials;
+using namespace ::chip::DeviceManager;
+using namespace ::chip::DeviceLayer;
+
+const char * TAG = "ota-provider-app";
+
+static DeviceCallbacks EchoCallbacks;
+
+// TODO: this should probably be done dynamically
+constexpr chip::EndpointId kOtaProviderEndpoint = 0;
+
+constexpr uint32_t kMaxBdxBlockSize                 = 1024;
+constexpr chip::System::Clock::Timeout kBdxTimeout  = chip::System::Clock::Seconds16(5 * 60); // Specification mandates >= 5 minutes
+constexpr chip::System::Clock::Timeout kBdxPollFreq = chip::System::Clock::Milliseconds32(500);
+const char * gOtaFilename                           = "hello-world.bin";
+BdxOtaSender bdxServer;
+
+const esp_partition_t * otaPartition = nullptr;
+uint32_t otaImageLen                 = 0;
+uint32_t otaTransferInProgress       = false;
+
+CHIP_ERROR OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset);
+void OnTransferComplete(void * context);
+void OnTransferFailed(void * context, BdxSenderErrorTypes status);
+
+chip::Callback::Callback<OnBdxBlockQuery> mOnBlockQuery(OnBlockQuery, nullptr);
+chip::Callback::Callback<OnBdxTransferComplete> mOnTransferComplete(OnTransferComplete, nullptr);
+chip::Callback::Callback<OnBdxTransferFailed> mOnTransferFailed(OnTransferFailed, nullptr);
+
+CHIP_ERROR OnBlockQuery(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof, uint32_t offset)
+{
+    if (otaTransferInProgress == false)
+    {
+        if (otaPartition == nullptr || otaImageLen == 0)
+        {
+            ESP_LOGE(TAG, "OTA partition not found");
+            return CHIP_ERROR_OPEN_FAILED;
+        }
+        otaTransferInProgress = true;
+    }
+
+    uint16_t blockBufAvailableLength = blockBuf->AvailableDataLength();
+    uint16_t transferBlockSize       = bdxServer.GetTransferBlockSize();
+
+    size = (blockBufAvailableLength < transferBlockSize) ? blockBufAvailableLength : transferBlockSize;
+
+    if (offset + size >= otaImageLen)
```

Comment:
One thing I just realized: where is `offset` coming from here?  Is it under attacker control?  As in, could the other side of this conversation cause `offset + size` to overflow here and this test to pass even though `offset` is very large?

If overflow is not a worry, please document why not.

Reply:
`offset` is not coming from the sender, it is locally tracked in `BdxOtaSender` class. There are two types of message to query a block: `BlockQuery` and `BlockQueryWithSkip` so, to handle both cases in a single callback I used offset which is managed by the `BdxOtaSender`.

Also, `esp_partition_read()` API returns error if offset is not in the range.

I've added a comment in the code as well.